{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_FOLDER = './Data' # Use the data folder provided in Tutorial 02 - Intro to Pandas.\n",
    "\n",
    "import glob\n",
    "import natsort\n",
    "import pandas as pd\n",
    "from dateutil.parser import parse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data field names\n",
    "Column names and the names of the variables we want to extract slightly vary between countries. Hence, first we define the column names of the _date_ column, _variable_ and _national total_ columns as well as the names of the fields for each country."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your answer here\n",
    "# Dataframe importing \n",
    "# Colums: country, month, averagy, new cases, death\n",
    "\n",
    "# column names and interesting variables for each country\n",
    "col_date = {\"guinea\" : \"Date\",\n",
    "            \"liberia\": \"Date\",\n",
    "            \"sl\" : \"date\"} # column name of the date column\n",
    "\n",
    "col_name={\"guinea\" : \"Description\",\n",
    "          \"liberia\": \"Variable\",\n",
    "          \"sl\" : \"variable\"} # column name of the indicator variables\n",
    "\n",
    "# TODO total vs new deaths / cases\n",
    "col_total = {\"guinea\":\"Totals\",\n",
    "             \"liberia\":\"National\",\n",
    "             \"sl\": \"National\"} # column name where the national totals per day are stored\n",
    "# TODO sum counties if no national data availble\n",
    "# TODO Make sure you handle all the different expressions for new cases and deaths that are used in the reports.\n",
    "col_counties={\"guinea\":[\"Conakry\",\"Gueckedou\",\"Macenta\",\"Dabola\",\"Kissidougou\",\"Dinguiraye\",\"Telimele\",\"Boffa\",\"Kouroussa\",\"Siguiri\",\"Pita\",\"Nzerekore\",\"Yomou\",\"Dubreka\",\"Forecariah\",\"Kerouane\",\"Coyah\",\"Dalaba\",\"Beyla\",\"Kindia\"],\n",
    "              \"liberia\":[\"\"],\n",
    "              \"sl\":[\"\"]}\n",
    "\n",
    "var_death =  {\"guinea\" : [\"New deaths registered today (confirmed)\",\"New deaths registered\"],\n",
    "             \"liberia\": [\"Newly reported deaths\"],# \"Total death/s in confirmed cases\"\n",
    "             \"sl\":[\"etc_new_deaths\"]}\n",
    "\n",
    "var_death_total =  {\"guinea\" : [\"Total deaths of confirmed\"],\n",
    "             \"liberia\": [\"Total death/s in confirmed cases\"],# \"Total death/s in confirmed cases\"\n",
    "             \"sl\":[\"death_confirmed\"]}\n",
    "\n",
    "var_cases = {\"guinea\" : [\"New cases of confirmed\"],\n",
    "             \"liberia\": [\"New case/s (confirmed)\"],\n",
    "             \"sl\":[\"cum_confirmed\"]}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Aggregate data frames\n",
    "Now, we loop over all files and import them, applying necessary clean-up operations such as type conversions, comma stripping etc. For each file (day), we extract one row with the corresponding values and merge them in `df`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Cases</th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>Total Deaths</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Country</th>\n",
       "      <th>Date</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">guinea</th>\n",
       "      <th>2014-08-04</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-26</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-27</th>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-30</th>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-31</th>\n",
       "      <td>29</td>\n",
       "      <td>3</td>\n",
       "      <td>337</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Cases  New Deaths  Total Deaths\n",
       "Country Date                                       \n",
       "guinea  2014-08-04      4           2           228\n",
       "        2014-08-26     10           5           292\n",
       "        2014-08-27     10           2           294\n",
       "        2014-08-30      9           5           324\n",
       "        2014-08-31     29           3           337"
      ]
     },
     "execution_count": 324,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs = [] # dataframes for each day\n",
    "for c in ['guinea','liberia','sl']:\n",
    "    for file in sorted(glob.glob(DATA_FOLDER+\"/ebola/\"+c+\"_data/*.csv\")):\n",
    "        df = pd.read_csv(file,index_col=None)\n",
    "        df = df.fillna(0) # Fill NA values\n",
    "        \n",
    "        # Get Data\n",
    "        # Dates\n",
    "        date = pd.to_datetime(df[1:2][col_date[c]]) # dates\n",
    "        \n",
    "        # New cases\n",
    "        df_cases = df[df[col_name[c]].isin(var_cases[c])] # extract only relevant rows\n",
    "        df_cases = df_cases[[col_date[c],col_name[c],col_total[c]]] # extract only relevant columns (date - name - total)\n",
    "        df_cases[0:1][col_total[c]].values[0] = str(df_cases[0:1][col_total[c]].values[0]).replace(',','') # strip commas for 1000s\n",
    "        df_cases[col_total[c]] = df_cases[col_total[c]].astype(int)     # convert to int (e.g., 10.0 or 10,0)\n",
    "        cases = int(df_cases[0:1][col_total[c]]) # cases\n",
    "\n",
    "        # New deaths\n",
    "        df_deaths = df[df[col_name[c]].isin(var_death[c])] # extract only relevant rows\n",
    "        df_deaths = df_deaths[[col_date[c],col_name[c],col_total[c]]] # extract only relevant columns (date - name - total)\n",
    "        deaths = int(df_deaths[0:1][col_total[c]]) # deaths \n",
    "        \n",
    "        # TODO Total Deaths\n",
    "        df_deaths_total = df[df[col_name[c]].isin(var_death_total[c])] # extract only relevant rows\n",
    "        df_deaths_total = df_deaths_total[[col_date[c],col_name[c],col_total[c]]] # extract only relevant columns (date - name - total)\n",
    "        deaths_total = int(df_deaths_total[0:1][col_total[c]]) # deaths \n",
    "        # TODO Total Deaths by County\n",
    "        \n",
    "        # new dataframe from data (one row)\n",
    "        df = pd.DataFrame({\"Country\":c,\"Date\":date,\"Cases\":cases,\"New Deaths\":deaths,\"Total Deaths\":deaths_total},index=None)\n",
    "        dfs.append(df)\n",
    "        \n",
    "for i in range(len(dfs)-1): # concatenate all dataframes (rows)\n",
    "    dfs[0] = pd.concat([dfs[0],dfs[i+1]])\n",
    "df=dfs[0]\n",
    "\n",
    "df=df.sort_values(by=['Country','Date']) # sort by country and date\n",
    "df=df.set_index(['Country','Date']) # new index\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data checking\n",
    "\n",
    "Let's check the data whether the index is unique and whether there are still NaN values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index unique:\n",
      "True\n",
      "New Deaths:\n",
      "0\n",
      "Total Deaths:\n",
      "0\n",
      "Cases:\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "# Data checking\n",
    "print(\"Index unique:\")\n",
    "print(df.index.is_unique) # True\n",
    "for field in (\"New Deaths\",\"Total Deaths\",\"Cases\"):\n",
    "    print(field+\":\")\n",
    "    print(df[df[field].isnull()].size) # get NaN colums (all 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Aggregate by month\n",
    "\n",
    "Now, we can regroup the data by month and sums each column by the month. Then, we rename the date column as month column "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Cases</th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>Total Deaths</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Country</th>\n",
       "      <th>Month</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">guinea</th>\n",
       "      <th>8</th>\n",
       "      <td>10</td>\n",
       "      <td>62</td>\n",
       "      <td>17</td>\n",
       "      <td>1475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>200</td>\n",
       "      <td>208</td>\n",
       "      <td>57</td>\n",
       "      <td>7217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>21</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">liberia</th>\n",
       "      <th>6</th>\n",
       "      <td>175</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>374</td>\n",
       "      <td>20</td>\n",
       "      <td>47</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>396</td>\n",
       "      <td>49</td>\n",
       "      <td>209</td>\n",
       "      <td>1762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1452</td>\n",
       "      <td>148</td>\n",
       "      <td>865</td>\n",
       "      <td>14562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2125</td>\n",
       "      <td>34</td>\n",
       "      <td>701</td>\n",
       "      <td>6188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1575</td>\n",
       "      <td>39</td>\n",
       "      <td>202</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1053</td>\n",
       "      <td>17355</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">sl</th>\n",
       "      <th>8</th>\n",
       "      <td>2630</td>\n",
       "      <td>15565</td>\n",
       "      <td>0</td>\n",
       "      <td>5877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4524</td>\n",
       "      <td>43555</td>\n",
       "      <td>8</td>\n",
       "      <td>12856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>5166</td>\n",
       "      <td>86149</td>\n",
       "      <td>99</td>\n",
       "      <td>25353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4389</td>\n",
       "      <td>94092</td>\n",
       "      <td>12</td>\n",
       "      <td>23581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1110</td>\n",
       "      <td>31463</td>\n",
       "      <td>11</td>\n",
       "      <td>6574</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               index  Cases  New Deaths  Total Deaths\n",
       "Country Month                                        \n",
       "guinea  8         10     62          17          1475\n",
       "        9        200    208          57          7217\n",
       "        10        21      6          15           562\n",
       "liberia 6        175     15          14           122\n",
       "        7        374     20          47           476\n",
       "        8        396     49         209          1762\n",
       "        9       1452    148         865         14562\n",
       "        10      2125     34         701          6188\n",
       "        11      1575     39         202             0\n",
       "        12      1053  17355           0             0\n",
       "sl      8       2630  15565           0          5877\n",
       "        9       4524  43555           8         12856\n",
       "        10      5166  86149          99         25353\n",
       "        11      4389  94092          12         23581\n",
       "        12      1110  31463          11          6574"
      ]
     },
     "execution_count": 327,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aggregate by month\n",
    "df=df.reset_index()\n",
    "df_month = pd.DataFrame(df.groupby([df.Country,df.Date.dt.month]).sum()).reset_index()\n",
    "df_month = df_month.rename(columns={'Date':'Month'})\n",
    "df_month = df_month.set_index(['Country','Month'])\n",
    "df_month"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BARCODE</th>\n",
       "      <th>Species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">MID1</th>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                         GROUP  \\\n",
       "BARCODE Species                                                                  \n",
       "MID1    Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  EXTRACTION CONTROL   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  EXTRACTION CONTROL   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  EXTRACTION CONTROL   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  EXTRACTION CONTROL   \n",
       "\n",
       "                                                             SAMPLE  Count  \n",
       "BARCODE Species                                                             \n",
       "MID1    Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown      7  \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown      2  \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  unknown      3  \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknown      3  \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown      7  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Write your answer here\n",
    "\n",
    "# Read in Excel files, concatenate them and add file name (MID1...9) for later merging\n",
    "i=1\n",
    "MID = pd.read_excel(DATA_FOLDER+'/microbiome/MID'+str(i)+'.xls', sheetname='Sheet 1', header=None,index_col=0)\n",
    "MID.columns=['Count']\n",
    "MID['BARCODE'] = \"MID1\" # add column with file name\n",
    "for i in range(2,10):\n",
    "    MID_tmp=pd.read_excel(DATA_FOLDER+'/microbiome/MID'+str(i)+'.xls', sheetname='Sheet 1', header=None,index_col=0)\n",
    "    MID_tmp.columns=['Count']\n",
    "    MID_tmp['BARCODE'] = \"MID\"+str(i)  # add column with file name\n",
    "    MID=pd.concat([MID, MID_tmp], axis=0)\n",
    "# read metadata\n",
    "\n",
    "MID.index.name=\"Species\"\n",
    "MID.head()\n",
    "metadata = pd.read_excel(DATA_FOLDER+'/microbiome/metadata.xls', sheetname='Sheet1',index_col=0)\n",
    "\n",
    "# merge metadata and files\n",
    "MID_merged = pd.merge(metadata, MID, left_index=True, right_on='BARCODE')\n",
    "# print(MID_merged.index.is_unique) # True\n",
    "MID_merged=MID_merged.fillna(\"unknown\")\n",
    "MID_merged=MID_merged.reset_index().set_index(['BARCODE','Species']) # new index\n",
    "MID_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
