{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 470,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_FOLDER = './Data' # Use the data folder provided in Tutorial 02 - Intro to Pandas.\n",
    "\n",
    "import glob\n",
    "import natsort\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from dateutil.parser import parse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data field names\n",
    "Column names and the names of the variables we want to extract slightly vary between countries. Hence, first we define the column names of the _date_ column, _variable_ and _national total_ columns as well as the names of the fields for each country."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 471,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dataframe importing \n",
    "# Colums: country, month, averagy, new cases, death\n",
    "\n",
    "# column names and interesting variables for each country\n",
    "col_date = {\"guinea\" : \"Date\",\n",
    "            \"liberia\": \"Date\",\n",
    "            \"sl\" : \"date\"} # column name of the date column\n",
    "\n",
    "col_name={\"guinea\" : \"Description\",\n",
    "          \"liberia\": \"Variable\",\n",
    "          \"sl\" : \"variable\"} # column name of the indicator variables\n",
    "\n",
    "col_total = {\"guinea\":\"Totals\",\n",
    "             \"liberia\":\"National\",\n",
    "             \"sl\": \"National\"} # column name where the national totals per day are stored\n",
    "\n",
    "col_counties={\"guinea\":[\"Conakry\",\"Gueckedou\",\"Macenta\",\"Dabola\",\"Kissidougou\",\"Dinguiraye\",\n",
    "                        \"Telimele\",\"Boffa\",\"Kouroussa\",\"Siguiri\",\"Pita\",\"Nzerekore\",\"Yomou\",\n",
    "                        \"Dubreka\",\"Forecariah\",\"Kerouane\",\"Coyah\",\"Dalaba\",\"Beyla\",\"Kindia\"],\n",
    "              \"liberia\":[\"Bomi County\",\"Bong County\",\"Grand Kru\",\"Lofa County\",\"Margibi County\",\n",
    "                         \"Maryland County\",\"Montserrado County\",\"Nimba County\",\"River Gee County\",\n",
    "                         \"RiverCess County\",\"Sinoe County\"],\n",
    "              \"sl\":[\"Kailahun\",\"Kenema\",\"Kono\",\"Kambia\",\"Koinadugu\",\"Bombali\",\"Tonkolili\",\"Port Loko\",\n",
    "                    \"Pujehun\",\"Bo\",\"Moyamba\",\"Bonthe\",\"Western area urban\",\"Western area rural\"]}\n",
    "\n",
    "var_death_new =  {\"guinea\" : [\"New deaths registered today (confirmed)\",\"New deaths registered\"],\n",
    "             \"liberia\": [\"Newly reported deaths\"],# \"Total death/s in confirmed cases\"\n",
    "             \"sl\":[\"etc_new_deaths\"]}\n",
    "\n",
    "var_death_total =  {\"guinea\" : [\"Total deaths of confirmed\"],\n",
    "             \"liberia\": [\"Total death/s in confirmed cases\"],# \"Total death/s in confirmed cases\"\n",
    "             \"sl\":[\"death_confirmed\"]}\n",
    "\n",
    "var_cases_new = {\"guinea\" : [\"New cases of confirmed\"],\n",
    "             \"liberia\": [\"New case/s (confirmed)\"],\n",
    "             \"sl\":[\"new_confirmed\"]}\n",
    "\n",
    "var_cases_tot = {\"guinea\" : [\"Total cases of confirmed\"],\n",
    "             \"liberia\": [\"Total confirmed cases\"],\n",
    "             \"sl\":[\"cum_confirmed\"]}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Aggregate data frames\n",
    "Now, we loop over all files and import them, applying necessary clean-up operations such as type conversions, comma stripping etc. For each file (day), we extract one row with the corresponding values and merge them in `df`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 472,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Cases</th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>Total Cases</th>\n",
       "      <th>Total Deaths</th>\n",
       "      <th>Total Deaths Counties</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Country</th>\n",
       "      <th>Date</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">guinea</th>\n",
       "      <th>2014-08-04</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>351</td>\n",
       "      <td>228</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-26</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>490</td>\n",
       "      <td>292</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-27</th>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>499</td>\n",
       "      <td>294</td>\n",
       "      <td>267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-30</th>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>533</td>\n",
       "      <td>324</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-31</th>\n",
       "      <td>29</td>\n",
       "      <td>3</td>\n",
       "      <td>563</td>\n",
       "      <td>337</td>\n",
       "      <td>337</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    New Cases  New Deaths  Total Cases  Total Deaths  \\\n",
       "Country Date                                                           \n",
       "guinea  2014-08-04          4           2          351           228   \n",
       "        2014-08-26         10           5          490           292   \n",
       "        2014-08-27         10           2          499           294   \n",
       "        2014-08-30          9           5          533           324   \n",
       "        2014-08-31         29           3          563           337   \n",
       "\n",
       "                    Total Deaths Counties  \n",
       "Country Date                               \n",
       "guinea  2014-08-04                    228  \n",
       "        2014-08-26                    292  \n",
       "        2014-08-27                    267  \n",
       "        2014-08-30                    324  \n",
       "        2014-08-31                    337  "
      ]
     },
     "execution_count": 472,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs = [] # dataframes for each day\n",
    "for c in ['guinea','liberia','sl']: #\n",
    "    for file in sorted(glob.glob(DATA_FOLDER+\"/ebola/\"+c+\"_data/*.csv\")):\n",
    "        df = pd.read_csv(file,index_col=None)\n",
    "        df = df.fillna(0) # Fill NA values\n",
    "        \n",
    "        # Get Data\n",
    "        columns=[col_date[c],col_name[c],col_total[c]]\n",
    "        \n",
    "        # Dates\n",
    "        date = pd.to_datetime(df[1:2][col_date[c]]) # dates\n",
    "        \n",
    "        # New cases\n",
    "        df_cases_new = df[df[col_name[c]].isin(var_cases_new[c])] # extract only relevant rows\n",
    "        df_cases_new = df_cases_new[columns] # extract only relevant columns (date - name - total)\n",
    "        df_cases_new[0:1][col_total[c]].values[0] = str(df_cases_new[0:1][col_total[c]].values[0]).replace(',','') # strip commas for 1000s\n",
    "        df_cases_new[col_total[c]] = df_cases_new[col_total[c]].astype(int)     # convert to int (e.g., 10.0 or 10,0)\n",
    "        cases_new = int(df_cases_new[0:1][col_total[c]]) # cases\n",
    "        \n",
    "        # Total cases\n",
    "        df_cases_total = df[df[col_name[c]].isin(var_cases_tot[c])] # extract only relevant rows\n",
    "        df_cases_total = df_cases_total[columns] # extract only relevant columns (date - name - total)\n",
    "        df_cases_total[0:1][col_total[c]].values[0] = str(df_cases_total[0:1][col_total[c]].values[0]).replace(',','') # strip commas for 1000s\n",
    "        df_cases_total[col_total[c]] = df_cases_total[col_total[c]].astype(int)     # convert to int (e.g., 10.0 or 10,0)\n",
    "        cases_total = int(df_cases_total[0:1][col_total[c]]) # cases\n",
    "\n",
    "        # New deaths\n",
    "        df_deaths_new = df[df[col_name[c]].isin(var_death_new[c])] # extract only relevant rows\n",
    "        df_deaths_new = df_deaths_new[columns] # extract only relevant columns (date - name - total)\n",
    "        deaths_new = int(df_deaths_new[0:1][col_total[c]]) # deaths \n",
    "        \n",
    "        # Total Deaths\n",
    "        df_deaths_total = df[df[col_name[c]].isin(var_death_total[c])] # extract only relevant rows\n",
    "        df_deaths_total = df_deaths_total[columns] # extract only relevant columns (date - name - total)\n",
    "        deaths_total = int(df_deaths_total[0:1][col_total[c]]) # deaths \n",
    "        \n",
    "        # Total Deaths by County\n",
    "        df_deaths_total_county_sum = df[df[col_name[c]].isin(var_death_total[c])] # extract only relevant rows\n",
    "        cols=list(set(df.columns.values).difference(columns))\n",
    "        df_deaths_total_county_sum = df_deaths_total_county_sum[cols] # extract only relevant columns (date - name - total)\n",
    "        for col in df_deaths_total_county_sum.columns:\n",
    "            df_deaths_total_county_sum[col]=df_deaths_total_county_sum[col].astype(int)\n",
    "        deaths_total_county_sum = np.nansum(df_deaths_total_county_sum.values)\n",
    "        \n",
    "        # new dataframe from data (one row per date)\n",
    "        df = pd.DataFrame({\"Country\":c,\"Date\":date,\"New Cases\":cases_new,\"Total Cases\":cases_total,\"New Deaths\":deaths_new,\"Total Deaths\":deaths_total,\"Total Deaths Counties\":deaths_total_county_sum},index=None)\n",
    "        dfs.append(df)\n",
    "        \n",
    "for i in range(len(dfs)-1): # concatenate all dataframes (rows)\n",
    "    dfs[0] = pd.concat([dfs[0],dfs[i+1]])\n",
    "df=dfs[0]\n",
    "\n",
    "df=df.sort_values(by=['Country','Date']) # sort by country and date\n",
    "df=df.set_index(['Country','Date']) # new index\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There seems to be a discrepancy between the new deaths (confirmed) and the total deaths (confirmed), as the new cases /deaths don't add up to the cumsum of the total cases / deaths of the next date. Maybe the data doesn't come from the same source. Also, the total deaths don't always quite add up with the sum of deaths per county, which may be due to incomplete data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 473,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>New Cases</th>\n",
       "      <th>New Deaths</th>\n",
       "      <th>Total Cases</th>\n",
       "      <th>Total Deaths</th>\n",
       "      <th>Total Deaths Counties</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Country</th>\n",
       "      <th>Date</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">sl</th>\n",
       "      <th>2014-12-01</th>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "      <td>5978</td>\n",
       "      <td>1549</td>\n",
       "      <td>1549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-12-04</th>\n",
       "      <td>37</td>\n",
       "      <td>3</td>\n",
       "      <td>6238</td>\n",
       "      <td>1648</td>\n",
       "      <td>1648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-12-05</th>\n",
       "      <td>54</td>\n",
       "      <td>8</td>\n",
       "      <td>6292</td>\n",
       "      <td>1669</td>\n",
       "      <td>1669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-12-06</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6317</td>\n",
       "      <td>1708</td>\n",
       "      <td>1708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-12-13</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6638</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    New Cases  New Deaths  Total Cases  Total Deaths  \\\n",
       "Country Date                                                           \n",
       "sl      2014-12-01         72           0         5978          1549   \n",
       "        2014-12-04         37           3         6238          1648   \n",
       "        2014-12-05         54           8         6292          1669   \n",
       "        2014-12-06          0           0         6317          1708   \n",
       "        2014-12-13          0           0         6638             0   \n",
       "\n",
       "                    Total Deaths Counties  \n",
       "Country Date                               \n",
       "sl      2014-12-01                   1549  \n",
       "        2014-12-04                   1648  \n",
       "        2014-12-05                   1669  \n",
       "        2014-12-06                   1708  \n",
       "        2014-12-13                      0  "
      ]
     },
     "execution_count": 473,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data checking\n",
    "\n",
    "Let's check the data whether the index is unique and whether there are still NaN values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 474,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index unique:\n",
      "True\n",
      "New Cases:\n",
      "0\n",
      "New Deaths:\n",
      "0\n",
      "Total Cases:\n",
      "0\n",
      "Total Deaths:\n",
      "0\n",
      "Total Deaths Counties:\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "# Data checking\n",
    "print(\"Index unique:\")\n",
    "print(df.index.is_unique) # True\n",
    "for col in df.columns:\n",
    "    print(col+\":\")\n",
    "    print(df[df[col].isnull()].size) # get NaN colums (all 0)\n",
    "    \n",
    "df=df.reset_index() # for next part"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Aggregate by month\n",
    "\n",
    "Now, we can regroup the data by month and sums each column by the month. Then, we rename the date column as month column "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 460,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Mean New Cases</th>\n",
       "      <th>Mean New Deaths</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Country</th>\n",
       "      <th>Month</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">guinea</th>\n",
       "      <th>8</th>\n",
       "      <td>12.40</td>\n",
       "      <td>3.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>13.00</td>\n",
       "      <td>3.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>6.00</td>\n",
       "      <td>15.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">liberia</th>\n",
       "      <th>6</th>\n",
       "      <td>2.14</td>\n",
       "      <td>2.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1.82</td>\n",
       "      <td>4.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>5.44</td>\n",
       "      <td>23.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6.17</td>\n",
       "      <td>36.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.36</td>\n",
       "      <td>28.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2.60</td>\n",
       "      <td>13.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1928.33</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">sl</th>\n",
       "      <th>8</th>\n",
       "      <td>18.55</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>34.41</td>\n",
       "      <td>0.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>56.71</td>\n",
       "      <td>3.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>59.90</td>\n",
       "      <td>0.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>32.60</td>\n",
       "      <td>2.20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Mean New Cases  Mean New Deaths\n",
       "Country Month                                 \n",
       "guinea  8               12.40             3.40\n",
       "        9               13.00             3.56\n",
       "        10               6.00            15.00\n",
       "liberia 6                2.14             2.00\n",
       "        7                1.82             4.27\n",
       "        8                5.44            23.22\n",
       "        9                6.17            36.04\n",
       "        10               1.36            28.04\n",
       "        11               2.60            13.47\n",
       "        12            1928.33             0.00\n",
       "sl      8               18.55             0.00\n",
       "        9               34.41             0.28\n",
       "        10              56.71             3.54\n",
       "        11              59.90             0.57\n",
       "        12              32.60             2.20"
      ]
     },
     "execution_count": 460,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Aggregate by month\n",
    "\n",
    "df_month = pd.DataFrame(df[[\"Country\",\"Date\",\"New Cases\",\"New Deaths\"]].groupby([df.Country,df.Date.dt.month]).mean()).reset_index()\n",
    "df_month.columns=[\"Country\",\"Month\",\"Mean New Cases\",\"Mean New Deaths\"]\n",
    "df_month = df_month.set_index(['Country','Month'])\n",
    "df_month.round(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It seems like Ebola peaked around September in Guinea, around August-October in Liberia and around October-November in Sierra Leone.\n",
    "\n",
    "_suspicious_: no deaths in November and December in Liberia but many new cases. Maybe erroneous data or breakout of a new epidemia."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DataFrame Merging\n",
    "First we merge all data frames:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 484,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>BARCODE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>7</td>\n",
       "      <td>MID1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <td>2</td>\n",
       "      <td>MID1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <td>3</td>\n",
       "      <td>MID1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <td>3</td>\n",
       "      <td>MID1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <td>7</td>\n",
       "      <td>MID1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Count BARCODE\n",
       "Species                                                          \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      7    MID1\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      2    MID1\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...      3    MID1\n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...      3    MID1\n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...      7    MID1"
      ]
     },
     "execution_count": 484,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Write your answer here\n",
    "\n",
    "# Read in Excel files, concatenate them and add file name (MID1...9) for later merging\n",
    "dfs=[]\n",
    "for i in range(1,10):\n",
    "    MID_tmp=pd.read_excel(DATA_FOLDER+'/microbiome/MID'+str(i)+'.xls', sheetname='Sheet 1', header=None,index_col=0)\n",
    "    MID_tmp.columns=['Count']\n",
    "    MID_tmp['BARCODE'] = \"MID\"+str(i)  # add column with file name\n",
    "    # new dataframe from data\n",
    "    dfs.append(MID_tmp)\n",
    "        \n",
    "for i in range(len(dfs)-1): # concatenate all dataframes\n",
    "    dfs[0] = pd.concat([dfs[0],dfs[i+1]])\n",
    "\n",
    "MID=dfs[0]\n",
    "MID.index.name=\"Species\"\n",
    "MID.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metadata import and Merging\n",
    "We can now import the metadata and merge the files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 485,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BARCODE</th>\n",
       "      <th>Species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">MID1</th>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                         GROUP  \\\n",
       "BARCODE Species                                                                  \n",
       "MID1    Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  EXTRACTION CONTROL   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  EXTRACTION CONTROL   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  EXTRACTION CONTROL   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  EXTRACTION CONTROL   \n",
       "\n",
       "                                                             SAMPLE  Count  \n",
       "BARCODE Species                                                             \n",
       "MID1    Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown      7  \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown      2  \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  unknown      3  \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknown      3  \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown      7  "
      ]
     },
     "execution_count": 485,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read metadata\n",
    "metadata = pd.read_excel(DATA_FOLDER+'/microbiome/metadata.xls', sheetname='Sheet1',index_col=0)\n",
    "\n",
    "# merge metadata and files\n",
    "MID_merged = pd.merge(metadata, MID, left_index=True, right_on='BARCODE')\n",
    "# print(MID_merged.index.is_unique) # True\n",
    "MID_merged=MID_merged.fillna(\"unknown\")\n",
    "MID_merged=MID_merged.reset_index().set_index(['BARCODE','Species']) # new index\n",
    "MID_merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
